src/geo/transform.js

```diff
@@ -1597,30 +1598,35 @@ class Transform {
         // The default camera position might have been compensated by the active projection model.
         const mercPixelsPerMeter = mercatorZfromAltitude(1, this._center.lat) * this.worldSize;
         const pos = this._computeCameraPosition(mercPixelsPerMeter);
-
         const elevationAtCamera = elevation.getAtPointOrZero(new MercatorCoordinate(...pos));
+
         const minHeight = this._minimumHeightOverTerrain() * Math.cos(degToRad(this._maxPitch));
         const terrainElevation = this.pixelsPerMeter / this.worldSize * elevationAtCamera;
         const cameraHeight = this._camera.position[2] - terrainElevation;
 
-        if (cameraHeight < minHeight) {
+        // If camera is under terrain or dragging at unsafe distance from terrain, force camera position above terrain
+        if (cameraHeight <= 0 || (cameraHeight < minHeight && isDragging)) {
             const center = this.locationCoordinate(this._center, this._centerAltitude);
             const cameraToCenter = [center.x - pos[0], center.y - pos[1], center.z - pos[2]];
-            const prevDistToCamera = vec3.length(cameraToCenter);
 
+            const prevDistToCamera = vec3.length(cameraToCenter);
             // Adjust the camera vector so that the camera is placed above the terrain.
             // Distance between the camera and the center point is kept constant.
             cameraToCenter[2] -= (minHeight - cameraHeight) / this._pixelsPerMercatorPixel;
-
             const newDistToCamera = vec3.length(cameraToCenter);
+
             if (newDistToCamera === 0)
                 return;
 
             vec3.scale(cameraToCenter, cameraToCenter, prevDistToCamera / newDistToCamera * this._pixelsPerMercatorPixel);
-            this._camera.position = [center.x - cameraToCenter[0], center.y - cameraToCenter[1], center.z * this._pixelsPerMercatorPixel - cameraToCenter[2]];
 
-            this._camera.orientation = orientationFromFrame(cameraToCenter, this._camera.up());
+            const newPosition = [center.x - cameraToCenter[0], center.y - cameraToCenter[1], center.z * this._pixelsPerMercatorPixel - cameraToCenter[2]];
+            this._setCameraPosition(newPosition);
             this._updateStateFromCamera();
+
+        // Set camera as constrained to keep zoom at safe distance from terrain
+        } else if (cameraHeight < minHeight) {
+            this._isCameraConstrained = true;
```

Comment:
Nit: Would it be possible to make logic a bit more concise, for instance to check for `cameraHeight < minHeight` only once?